potoken.py

```diff
@@ -0,0 +1,255 @@
+            # handler is a function returning a tuple of status, headers, page text
+            '/404': lambda: ('404 Not Found', [("Content-Type", "text/plain")], 'Not Found'),
+            '/': lambda: ('302 Found', [('Location', '/token')], '/token'),
+            '/token': self.get_potoken,
+            '/update': self.request_update
+        }
+        return handlers.get(route) or handlers['/404']
+
+    def app(self, environ, start_response):
+        route = environ['PATH_INFO']
+
+        handler = self.get_route_handler(route)
+        status, headers, page = handler()
+
+        start_response(status, headers)
+        return [page.encode('utf8')]
+
+    def run(self):
+        logger.info(f'Starting web-server at {self.bind_address}:{self.port}')
+        self._httpd = make_server(self.bind_address, self.port, self.app, ThreadingWSGIServer)
+        with self._httpd:
+            self._httpd.serve_forever()
+
+    def stop(self):
+        if self._httpd is None:
+            return
+        self._httpd.shutdown()
+
+
+def main(update_interval, bind_address, port) -> None:
+    loop = nodriver.loop()
+
+    potoken_extractor = PotokenExtractor(loop, update_interfal=update_interval)
+    potoken_server = PotokenServer(potoken_extractor, port=port, bind_address=bind_address)
+
+    extractor_task = loop.create_task(potoken_extractor.run())
+    server_task = loop.create_task(asyncio.to_thread(potoken_server.run))
+
+    try:
+        main_task = asyncio.gather(extractor_task, server_task)
+        loop.run_until_complete(main_task)
+    except Exception:
+        # exceptions raised by the tasks are intentionally propogated
+        # to ensure process exit code is 1 on error
+        raise
+    except (KeyboardInterrupt, asyncio.CancelledError):
+        logger.info('Stopping...')
+    finally:
+        potoken_server.stop()
+
+
+def set_logging(log_level=logging.DEBUG):
+    log_format = '%(asctime)s.%(msecs)03d [%(name)s] [%(levelname)s] %(message)s'
+    datefmt = '%Y/%m/%d %H:%M:%S'
+    logging.basicConfig(level=log_level, format=log_format, datefmt=datefmt)
+    logging.getLogger('asyncio').setLevel(logging.INFO)
+    logging.getLogger('nodriver').setLevel(logging.WARNING)
+    logging.getLogger('uc').setLevel(logging.WARNING)
+    logging.getLogger('websockets').setLevel(logging.WARNING)
+
+
+def args_parse():
+    description = '''
+Retrieve potoken using Chromium runned by nodriver, serve it on a json endpoint
+
+    Token is generated on startup, and then every UPDATE_INTERVAL seconds.
+    With web-server running on default port, the token is available on the
+    http://127.0.0.1:8080/token endpoint. It is possible to request immediate
+    token regeneration by accessing http://127.0.0.1:8080/update
+    '''
+    parser = argparse.ArgumentParser(description=description, formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('--update-interval', '-u', type=int, default=3600 * 12,
```

Comment:
could you make it so it refreshes the token every hour by default? this is much better for privacy.

Reply:
Personally I would expect the backend to be smart enough to merge history from multiple tokens, considering neither IP nor hardware fingerprint is changing between updates. However a hour does seems to be a more reasonable number then twelve, considering no retry on failure is implemented.